vault/logical_system_namespaces_test.go

```diff
@@ -0,0 +1,413 @@
+		require.Equal(t, res.Data, map[string]interface{}{}, "list data has unexpected elements")
+	})
+
+	t.Run("list includes non-root namespaces", func(t *testing.T) {
+		testCreateNamespace(t, nil, b, "foo", nil)
+		testCreateNamespace(t, nil, b, "bar", nil)
+
+		req := logical.TestRequest(t, logical.ListOperation, "namespaces")
+		res, err := b.HandleRequest(rootCtx, req)
+
+		require.NoError(t, err)
+		require.NotEmpty(t, res.Data["keys"], "keys is empty")
+		require.NotEmpty(t, res.Data["key_info"], "key_info is empty")
+
+		keys, ok := res.Data["keys"].([]string)
+		require.True(t, ok, "keys is not a list")
+		keyInfo, ok := res.Data["key_info"].(map[string]interface{})
+		require.True(t, ok, "key_info is not a map")
+
+		require.Equal(t, len(keys), 2, "expected two entries in keys")
+		require.Equal(t, len(keyInfo), 2, "expected two entries in key_info")
+
+		for _, path := range keys {
+			info, ok := keyInfo[path]
+			require.True(t, ok, fmt.Sprintf("key_info does not have path %q which is present in keys", path))
+			var ns namespace.Namespace
+			require.NoError(t, mapstructure.Decode(info, &ns), "key_info entry is not a namespace")
+			require.Equal(t, ns.Path, path, "path in key does not match path in namespace struct")
+			require.NotEmpty(t, ns.ID, "namespace ID should not be empty")
+			require.NotEqual(t, ns.ID, namespace.RootNamespaceID, "list should not include root namespace")
+		}
+	})
+
+	t.Run("list only includes one level of namespaces", func(t *testing.T) {
+		fooNs := testCreateNamespace(t, nil, b, "foo", nil)
+		nestedCtx := namespace.ContextWithNamespace(rootCtx, fooNs)
+		testCreateNamespace(t, nil, b, "bar", nil)
+		testCreateNamespace(t, nestedCtx, b, "baz", nil)
+
+		req := logical.TestRequest(t, logical.ListOperation, "namespaces")
+		res, err := b.HandleRequest(rootCtx, req)
+
+		require.NoError(t, err)
+		require.NotEmpty(t, res.Data["keys"], "keys is empty")
+		require.NotEmpty(t, res.Data["key_info"], "key_info is empty")
+
+		keys, ok := res.Data["keys"].([]string)
+		require.True(t, ok, "keys is not a list")
+		keyInfo, ok := res.Data["key_info"].(map[string]interface{})
+		require.True(t, ok, "key_info is not a map")
+
+		require.Equal(t, 2, len(keys), "expected two entries in keys")
+		require.Equal(t, 2, len(keyInfo), "expected two entries in key_info")
+		require.Subset(t, keys, []string{"foo/", "bar/"})
+
+		for _, path := range keys {
+			info, ok := keyInfo[path]
+			require.True(t, ok, fmt.Sprintf("key_info does not have path %q which is present in keys", path))
+			var ns namespace.Namespace
+			require.NoError(t, mapstructure.Decode(info, &ns), "key_info entry is not a namespace")
+			require.Equal(t, ns.Path, path, "path in key does not match path in namespace struct")
+			require.NotEmpty(t, ns.ID, "namespace ID should not be empty")
+			require.NotEqual(t, ns.ID, namespace.RootNamespaceID, "list should not include root namespace")
+		}
+	})
+
+	t.Run("list nested namespaces", func(t *testing.T) {
```

Comment:
Test is currently failing due to ListNamespaces implementation, see https://github.com/openbao/openbao/pull/1041#discussion_r1975059934